src/rnnlm/arpa-sampling.h

```diff
@@ -0,0 +1,131 @@
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ARPA_SAMPLING_H_
+#define ARPA_SAMPLING_H_
+
+#include <algorithm>
+#include <map>
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <sstream>
+#include <vector>
+
+#include <math.h>
+#include <sys/time.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include "fst/fstlib.h"
+#include "util/common-utils.h"
+#include "lm/arpa-file-parser.h"
+
+namespace kaldi {
+
+class ArpaSampling : public ArpaFileParser {
+ public:
+  friend class ArpaSamplingTest;
+
+  // HistType represents a history
+  typedef std::vector<int32> HistType;
+  // WordToProbsMap represents the words and their probabilities given
+  // an arbitrary history
+  typedef std::unordered_map<int32, std::pair<BaseFloat, BaseFloat> > WordToProbsMap; 
+  // NgramType represents the map between a history and the map of existing words
+  // and their probabilities given this history
+  typedef std::unordered_map<HistType, WordToProbsMap, VectorHasher<int32> > NgramType;
+  // HistWeightsType represents the map between histories (with weights) and
+  // their computed weights
+  typedef std::unordered_map<HistType, BaseFloat, VectorHasher<int32> > HistWeightsType;
+
+  // constructor
+  ArpaSampling(ArpaParseOptions options, fst::SymbolTable* symbols)
+     : ArpaFileParser(options, symbols) { 
+       ngram_order_ = 0;
+       num_words_ = 0;
+       bos_symbol_ = "<s>";
+       eos_symbol_ = "</s>";
+       unk_symbol_ = "<unk>";
+  }
+  
+  // This function computes the unigram distribution of all vocab words which are
+  // represented by integers from 1 to num_words_ (0 is reserved for epsilon)
+  // Probabilities of integers not represented by any valid word is 0, e.g.
+  // 0 has probability 0.0
+  void GetUnigramDistribution(std::vector<BaseFloat> *unigram_probs);
+
+  // This function reads in a list of histories with input weights and returns
+  // 1) the non_unigram_probs (maps the higher-than-unigram words to their
+  // corresponding probabilities given the list of histories) and 
+  // 2) a scalar unigram_weight = sum of history_weight * backoff_weight of that history
+  BaseFloat GetDistribution(const std::vector<std::pair<HistType, BaseFloat> > 
+      &histories, std::unordered_map<int32, BaseFloat> *non_unigram_probs);
+ 
+ protected:
+  // ArpaFileParser overrides.
+  virtual void HeaderAvailable(); 
+  virtual void ConsumeNGram(const NGram& ngram);
+  virtual void ReadComplete() {}
+
+ private:
+  // This function returns the log probability of a LM state, [history word],
```

Comment:
There is no such thing as logprob of LM state.
It returns the log probability of a word 'word' given a history 'history'.
Please remove 'order' from the interface, as it's redundant with 'history'
(it has to be the same as history.size() + 1).   Make the function const.
Rename it to GetLogprob, as it returns the log-prob.